Recommendation request functional

SergeyZhitenev opened the pull request.

SergeyZhitenev committed:
feat: add recommendation request functional

Sanasar1 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,33 @@
+
+@Controller
+public class RecommendationRequestController {
+    private RecommendationRequestService recommendationRequestService;
```
Забыл заинджектить RecommendationRequestService

Sanasar1 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,105 @@
+
+    public RecommendationRequestDto create(RecommendationRequestDto recommendationRequestDto) {
+        RecommendationRequest recommendationRequest = mapToFullRecommendationRequest(recommendationRequestDto);
+        validateAbilityToSendRepeatRequest(recommendationRequest);
```
- Мне кажется, что лучше сначала провалидировать, и только потом спамить дтошку в сущность. Так логичнее на первом шаге проверять, все ок или нет и сразу кидать ошибку, если что-то не так, а не посреди бизнес логики. Если я что-то не разглядел и/или ты не согласен со мной, дай знать
- Непонятно, в чем разница между Ability и Skill. Кто-то может подумать, что это про одно и то же, а кто-то может подумать что skill - навык, ability - возможность

Sanasar1 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,105 @@
+        return recommendationRequestRepository.save(recommendationRequest);
+    }
+
+    private void validateAbilityToSendRepeatRequest(RecommendationRequest recommendationRequest) {
```
Я бы подумал, как можно поменять название. Что мне тут бросается в глаза:
- Та же история про skill и ability
- Два глагола в названии метода. Если не залесть в код, не вчитаться нормально в название метода, или просто плохо английский знать, то можно подумать, что метод и валидирует, и отправляет, что неправда

Sanasar1 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,105 @@
+        RecommendationRequest lastRecommendationRequest = recommendationRequestRepository.findLatestPendingRequest(requesterId, receiverId)
+                .orElse(null);
+        if (lastRecommendationRequest != null) {
+            boolean isAcceptable = ChronoUnit.MONTHS.between(LocalDateTime.now(), lastRecommendationRequest.getCreatedAt()) >= ACCEPTABLE_MONTH_BETWEEN_REQUEST;
```
Как альтернативу тут можно еще использовать Period. Это не замечание, просто для общего развития, если интересно посмотреть)
У Коли вчера подглядел, можешь тоже у него глянуть

Sanasar1 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,105 @@
+        long requesterId = recommendationRequest.getRequester().getId();
+        long receiverId = recommendationRequest.getReceiver().getId();
+        RecommendationRequest lastRecommendationRequest = recommendationRequestRepository.findLatestPendingRequest(requesterId, receiverId)
+                .orElse(null);
```
Давай сразу с помощью .orElseThrow() кинем тут ошибку, а не будем возвращать null:
- Возвращать null в принципе не очень хорошо
- Сделаем код чище, избавившись от проверки на null ниже

Sanasar1 commented on src/main/java/school/faang/user_service/service/UserService.java:
```diff
@@ -0,0 +1,17 @@
+
+    public User getUserById(long id) {
+        return userRepository.findById(id).orElseThrow(
+                () -> new IllegalArgumentException("User with id: %s not found".formatted(id)));
```
EntityNotFoundException или ResourceNotFoundException тут лучше подойдут. Они явно говорят в своем названии, что сущность/ресурс не найден. Лично я предпочитаю EntityNotFoundException (потому что мы пытаемся именно сущность достать), но мне подмечали и про ResourceNotFoundException
То же самое в skillService.getSkillById()

Sanasar1 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,105 @@
+                        ((subGenStream, stream) -> stream))
+                .distinct()
+                .toList();
+    }
```
Классно использовал стримы, особенно .reduce(), мне зашло

Sanasar1 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,105 @@
+                .orElseThrow(() -> new IllegalArgumentException("Unable to find recommendation request with id " + id));
+    }
+
+    public RecommendationRequest rejectRequest(long id, RejectionDto rejection) {
```
Нужно возвращать dto, контроллер никак не должен взаимодействовать с сущностями

Sanasar1 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,105 @@
+        return recommendationRequestMapper.toDto(recommendationRequest);
+    }
+
+    public List<RecommendationRequest> getRequests(RequestFilterDto filterDto) {
```
Нужно возвращать dto, контроллер никак не должен взаимодействовать с сущностями

Sanasar1 commented on src/main/java/school/faang/user_service/service/RecommendationRequestService.java:
```diff
@@ -0,0 +1,105 @@
+                .toList();
+    }
+
+    public RecommendationRequest getRequest(long id) {
```
Нужно возвращать dto, контроллер никак не должен взаимодействовать с сущностями

Sanasar1 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,33 @@
+
+
+@Controller
+public class RecommendationRequestController {
```
Нужно добавить `@Validated` на уровне класса. Если не сделать, то спринг в методах ничего не провалидирует. Как альтернативу можно просто над конкретными методами, где нужно валидация, поставить аннотацию. И тогда можно будет убрать `@Validated` из параметров методов

Sanasar1 commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,33 @@
+        return recommendationRequestService.getRequests(filter);
+    }
+
+    public RecommendationRequest getRecommendationRequest(long id) {
```
Я бы еще добавил проверку, что айдишник положительный

Sanasar1 commented on src/main/java/school/faang/user_service/dto/recommendation/RecommendationRequestDto.java:
```diff
@@ -0,0 +1,21 @@
+        Long id,
+        @NotBlank @Size(max = 4096) String message,
+        @NotNull RequestStatus status,
+        @NotNull List<SkillRequestDto> skills,
```
Может стоит только айдишнки передавать, а не всю дтошку? Сейчас выглядит безобидно, учитывая, что в SkillRequestDto только два поля. Но если мы надобавим туда еще больше полей, и еще хуже, если там окажутся другие дто'шки, то это много лишней информации, которую мы безпричинно тоскаем по интернету

SergeyZhitenev committed:
fix: возвращаемые параметры в классе Controller
Обновлен метод валидации в RecommendationRequestService
Изменен exception в UserService и SkillService

ipasynkov-epam commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,40 @@
+public class RecommendationRequestController {
+    private final RecommendationRequestService recommendationRequestService;
+
+    @Validated
```
Можно поместить на уровень класса.
Либо, если у нас в метод контроллера приходит объект, то использовать
@Valid @RequestBody ...dto

ipasynkov-epam commented on src/main/java/school/faang/user_service/controller/recommendation/RecommendationRequestController.java:
```diff
@@ -0,0 +1,40 @@
+        return recommendationRequestService.create(recommendationRequest);
+    }
+
+    @Validated
```
Если это контроллер, то должны быть маппинги для ендпоинтов?
типа @PostMapping("/path") ?

ipasynkov-epam commented on src/main/java/school/faang/user_service/filters/recommendationRequestFilters/Filter.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.util.stream.Stream;
+
+public interface Filter<T, F> {
```
не нужно делать дженерик:
1) в методы isApplicable и apply во всех случаях приходит RequestFilterDto
2) Другие фильтры, пусть просто наследуются от текущего класса Filter

ipasynkov-epam reviewed (commented):
хорошо. теперь еще нужны тесты)

SergeyZhitenev committed:
Merge branch 'refs/heads/Wyvern-master-stream7' into feature-BJS2-40460

# Conflicts:
#	src/main/java/school/faang/user_service/service/SkillService.java
#	src/main/java/school/faang/user_service/service/UserService.java

SergeyZhitenev committed:
Merge branch 'refs/heads/Wyvern-master-stream7' into feature-BJS2-40460

SergeyZhitenev committed:
feat: добавлены тесты
рефакторинг RecommendationRequestService
Изменен RecommendationRequestDto

SergeyZhitenev committed:
feat: добавленны mapping аннотации

SergeyZhitenev committed:
fix: исправлен тест

SergeyZhitenev committed:
feat: добавлена RequestBody аннотация в RecommendationRequestController
